### PR TITLE
gh: add module

### PR DESCRIPTION
**Pull request by Gerschtli**
### Description

Add module for new GitHub CLI Tool.

**Note**: I ran all tests but like in master the evaluation failed but my added test cases work.
```
$ nix-shell --pure tests -A run.all
trace: warning: In file /home/tobias/projects/home-manager/tests/lib/types/list-or-dag-merge.nix
a list is being assigned to the option 'tested.dag'.
This will soon be an error due to the list form being deprecated.
Please use the attribute set form instead with DAG functions to
express the desired order of entries.

trace: warning: In file /home/tobias/projects/home-manager/tests/lib/types/list-or-dag-merge.nix
a list is being assigned to the option 'tested.dag'.
This will soon be an error due to the list form being deprecated.
Please use the attribute set form instead with DAG functions to
express the desired order of entries.

error: --- TypeError ------------------------------------------------------------------------------------------------------------------------ nix-shell
'renderBinding' at /home/tobias/projects/home-manager/modules/programs/ncmpcpp.nix:23:19 called with unexpected argument '_module'
(use '--show-trace' to show detailed location information)
```

### Checklist

- [x] Change is backwards compatible.

- [x] Code formatted with `./format`.

- [x] Code tested through `nix-shell --pure tests -A run.all`.

- [x] Test cases updated/added. See [example](https://github.com/rycee/home-manager/commit/f3fbb50b68df20da47f9b0def5607857fcc0d021#diff-b61a6d542f9036550ba9c401c80f00ef).

- [x] Commit messages are formatted like

    ```
    {component}: {description}

    {long description}
    ```

    See [CONTRIBUTING](https://github.com/rycee/home-manager/blob/master/doc/contributing.adoc#sec-commit-style) for more information and [recent commit messages](https://github.com/rycee/home-manager/commits/master) for examples.

- If this PR adds a new module

  - [x] Added myself as module maintainer. See [example](https://github.com/rycee/home-manager/blob/068ff76a10e95820f886ac46957edcff4e44621d/modules/programs/lesspipe.nix#L6).

  - [x] Added myself and the module files to `.github/CODEOWNERS`.

### CODE REVIEW

File: modules/programs/gh.nix [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+
+    programs.bash.initExtra = mkIf cfg.enableBashIntegration ''
+      eval "$(${package}/bin/gh completion --shell bash)"
+    '';
+
+    programs.zsh.initExtra = mkIf cfg.enableZshIntegration ''
+      eval "$(${package}/bin/gh completion --shell zsh)"
+    '';
+
+    programs.fish.shellInit = mkIf cfg.enableFishIntegration ''
```
Comment (teto) [inline modules/programs/gh.nix:84]: I don't think that's necessary, the package already installs those completion files (I have autocompletion just by virtue of installing gh).

**Comment by teto** [comment]
I've trouble seeing the added value of adding this in HM. The config looks straightforward and the options dont interact too much with the rest of the computer ?

File: modules/programs/gh.nix (hunk 1 above)
Comment (Gerschtli) [reply modules/programs/gh.nix]: Yes, you are right. Thank you

**Comment by Gerschtli** [comment]
@teto I don't understand what you mean. With this module you can configure aliases, gitProtocol and editor config values.

**Comment by rycee** [comment]
Thanks! Merged to master with some minor changes.

**Comment by bbigras** [comment]
Where do you put the github token if `~/.config/gh/config.yml` is read-only (handled by home-manager)?

```
❯ gh auth login
? What account do you want to log into? GitHub.com
- Logging into github.com
? How would you like to authenticate? Login with a web browser

! First copy your one-time code: <something>
- Press Enter to open github.com in your browser...
failed to authenticate via web browser: open /home/bbigras/.config/gh/config.yml: read-only file system
```

**Comment by Gerschtli** [comment]
IIRC, there is an environment variable called `GITHUB_TOKEN` which you can also use.

**Comment by bbigras** [comment]
Yeah, but I would prefer not having to set it every time or having this set for every cli commands I use.

I'm guessing I have to wait for them to support keychain or netrc (.netrc is not really better than `GITHUB_TOKEN`).

**Comment by jamesottaway** [comment]
I was just working through this, deleted my HM-managed `~/.config/gh/config.yml` and ran `gh auth login` to see what the CLI does.

Long story short, it writes another file: `~/.config/gh/hosts.yml` with the following format:

```yml
github.com:
    user: jamesottaway
    oauth_token: tokengoeshere
    git_protocol: ssh
```

I'm going to add `xdg.configFile."gh/hosts.yml"` to my config, but other than needing to manage the token carefully, I could see this module helping out.

File: modules/programs/gh.nix [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+        The protocol to use when performing Git operations.
+      '';
+    };
+  };
+
+  config = mkIf cfg.enable {
+    home.packages = [ pkgs.gitAndTools.gh ];
+
+    xdg.configFile."gh/config.yml".text =
+      builtins.toJSON { inherit (cfg) aliases editor gitProtocol; };
```
[the remark this answers is not in the document]
Comment (terlar) [reply modules/programs/gh.nix:51]: There is no built in function in Nix to write YAML, however YAML is a superset of JSON so it is fine to write JSON as YAML

**Pull request merged**